File: functions/runtime/src/Perper.WebJobs.Extensions/Model/Context.cs
using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Apache.Ignite.Core.Client;
using Microsoft.Extensions.Logging;
using Perper.WebJobs.Extensions.Cache;
using Perper.WebJobs.Extensions.Services;

namespace Perper.WebJobs.Extensions.Model
{
    public class Context : IContext
    {
        private readonly FabricService _fabric;
        private readonly IIgniteClient _ignite;
        private readonly PerperInstanceData _instance;
        private readonly PerperBinarySerializer _serializer;
        private readonly IState _state;
        private readonly ILogger<Context> _logger;

        public IAgent Agent => new Agent(_instance.Agent, _fabric.AgentDelegate, this, _serializer);

        public Context(FabricService fabric, PerperInstanceData instance, IIgniteClient ignite, PerperBinarySerializer serializer, IState state, ILogger<Context> logger)
        {
            _fabric = fabric;
            _ignite = ignite;
            _instance = instance;
            _serializer = serializer;
            _state = state;
            _logger = logger;
        }

        public async Task<(IAgent, TResult)> StartAgentAsync<TResult>(string delegateName, object? parameters = default)
        {
            var agentDelegate = delegateName;
            var callDelegate = delegateName;

            var agentName = GenerateName(agentDelegate);
            var agent = new Agent(agentName, agentDelegate, this, _serializer);

            var result = await agent.CallFunctionAsync<TResult>(callDelegate, parameters);

            return (agent, result);
        }

        public async Task<IStream<TItem>> StreamFunctionAsync<TItem>(string functionName, object? parameters = default, StreamFlags flags = StreamFlags.Default)
        {
            var streamName = GenerateName(functionName);
            await CreateStreamAsync(streamName, StreamDelegateType.Function, functionName, parameters, typeof(TItem), flags);
            return new Stream<TItem>(streamName, _instance, _fabric, _ignite, _serializer, _state, _logger);
        }

        public async Task<IStream> StreamActionAsync(string actionName, object? parameters = default, StreamFlags flags = StreamFlags.Default)
        {
            var streamName = GenerateName(actionName);
            await CreateStreamAsync(streamName, StreamDelegateType.Action, actionName, parameters, null, flags);
            return new Stream(streamName, _fabric, _ignite);
        }

        public IStream<TItem> DeclareStreamFunction<TItem>(string functionName)
        {
            var streamName = GenerateName(functionName);
            var stream = new Stream<TItem>(streamName, _instance, _fabric, _ignite, _serializer, _state, _logger);
            stream.FunctionName = functionName;
            return stream;
        }

        public async Task InitializeStreamFunctionAsync<TItem>(IStream<TItem> stream, object? parameters = default, StreamFlags flags = StreamFlags.Default)
        {
            var streamInstance = (Stream<TItem>)stream;
            if (streamInstance.FunctionName == null)
            {
                throw new InvalidOperationException("Stream is already initialized");
            }
            await CreateStreamAsync(streamInstance.StreamName, StreamDelegateType.Function, streamInstance.FunctionName!, parameters, typeof(TItem), flags);
            streamInstance.FunctionName = null;
        }

        public async Task<(IStream<TItem>, string)> CreateBlankStreamAsync<TItem>(StreamFlags flags = StreamFlags.Default)
        {
            var streamName = GenerateName();
            await CreateStreamAsync(streamName, StreamDelegateType.External, "", null, typeof(TItem), flags);
            return (new Stream<TItem>(streamName, _instance, _fabric, _ignite, _serializer, _state, _logger), streamName);
        }

        private async Task CreateStreamAsync(string streamName, StreamDelegateType delegateType, string delegateName, object? parameters, Type? type, StreamFlags flags)
        {
            var streamsCache = _ignite.GetCache<string, StreamData>("streams");
            await streamsCache.PutAsync(streamName, new StreamData
            {
                Agent = _instance.Agent,
                AgentDelegate = _fabric.AgentDelegate,
                Delegate = delegateName,
                DelegateType = delegateType,
                Parameters = parameters,
                Listeners = new List<StreamListener>(),
                IndexType = (flags & StreamFlags.Query) != 0 && type != null ? PerperTypeUtils.GetJavaTypeName(type) ?? type.Name : null,
                IndexFields = (flags & StreamFlags.Query) != 0 && type != null ? _serializer.GetQueriableFields(type) : null,
                Ephemeral = (flags & StreamFlags.Ephemeral) != 0,
                LastModified = DateTime.UtcNow
            });
        }

        public async Task<CallData> CallAsync(string agentName, string agentDelegate, string callDelegate, object? parameters)
        {
            var callsCache = _ignite.GetCache<string, CallData>("calls");
            var callName = GenerateName(callDelegate);
            await callsCache.PutAsync(callName, new CallData
            {
                Agent = agentName,
                AgentDelegate = agentDelegate,
                Delegate = callDelegate,
                CallerAgentDelegate = _fabric.AgentDelegate,
                Caller = _instance.InstanceName,
                Finished = false,
                LocalToData = true,
                Parameters = parameters,
            });

            var (key, notification) = await _fabric.GetCallNotification(callName);
            await _fabric.ConsumeNotification(key);

            var callResult = await callsCache.GetAndRemoveAsync(notification.Call);
            var call = callResult.Value;

            if (call.Error != null)
            {
                throw new Exception("Exception in call to '" + callDelegate + "': " + call.Error);
            }
            else
            {
                return call;
            }
        }

        private string GenerateName(string? baseName = null)
        {
            return $"{baseName}-{Guid.NewGuid()}";
        }
    }
}

File: functions/python/perper/model/context.py
from datetime import datetime
import uuid
import logging

from pyignite.datatypes import CollectionObject
from pyignite.utils import hashcode, entity_id
from perper.utils import PerperThinClient

from perper.services import PerperTypeUtils
from perper.services import Serializer
from perper.cache import CallData
from perper.cache import StreamData, ParameterData
from perper.cache import StreamDelegateType
from perper.model import StreamFlags
from perper.model import Stream
from perper.model import Agent


class Context:
    def __init__(
        self,
        instance,
        fabric,
        state,
        ignite=PerperThinClient(),
        logger=logging.getLogger("context"),
    ):
        self.ignite = ignite
        self.instance = instance
        self.serializer = Serializer()
        self.fabric = fabric
        self.state = state
        self._logger = logger
        self.agent = Agent(
            agent_name=self.instance.agent,
            agent_delegate=self.fabric.agent_delegate,
            context=self,
            serializer=self.serializer,
        )

    async def start_agent(self, delegate_name, parameters):
        agent_delegate = delegate_name
        call_delegate = delegate_name

        agent_name = self.generate_name(agent_delegate)
        agent = Agent(
            agent_name=agent_name,
            agent_delegate=agent_delegate,
            context=self,
            serializer=self.serializer,
        )

        result = await agent.call_function(call_delegate, parameters)

        return (agent, result)

    def stream_function(self, function_name, parameters, flags):
        stream_name = self.generate_name(function_name)
        self.create_stream(
            stream_name,
            StreamDelegateType.function,
            function_name,
            parameters,
            None,
            flags,
        )
        # return Stream(stream_name, self.fabric, self.ignite, instance = self.instance, serializer = self.serializer, state = self.state)
        return Stream(streamname=stream_name)

    def stream_action(self, action_name, parameters, flags):
        stream_name = self.generate_name(action_name)
        self.create_stream(
            stream_name, StreamDelegateType.action, action_name, parameters, None, flags
        )
        # return Stream(stream_name, self.fabric, self.ignite, instance = self.instance, serializer = self.serializer, state = self.state)
        return Stream(streamname=stream_name)

    def declare_stream_function(self, function_name):
        stream_name = self.generate_name(function_name)
        stream = Stream(
            stream_name,
            self.fabric,
            self.ignite,
            instance=self.instance,
            serializer=self.serializer,
            state=self.state,
        )
        stream.function_name = function_name
        return stream

    def initialize_stream_function(self, stream, parameters, flags):
        stream_instance = stream
        if stream_instance.function_name == None:
            raise Exception("Stream is already initialized")

        self.create_stream(
            stream_instance.stream_name,
            StreamDelegateType.function,
            stream_instance.function_name,
            parameters,
            None,
            flags,
        )

    def create_blank_stream(self, flags=StreamFlags.default):
        stream_name = self.generate_name()
        self.create_stream(
            stream_name, StreamDelegateType.external, "", {}, None, flags
        )

    def create_stream(
        self, stream_name, delegate_type, delegate_name, parameters, type_, flags
    ):
        streams_cache = self.ignite.get_cache("streams")
        index_type = None
        stream_data = StreamData(
            agent=self.instance.agent,
            agentdelegate=self.fabric.agent_delegate,
            delegate=delegate_name,
            delegatetype=(entity_id("StreamDelegateType"), delegate_type.value),
            # TODO: Parameters should be passed as BinaryObject
            parameters=ParameterData(parameters=(1, parameters)),
            listeners=(1, []),
            # Perper type utils is not implemented
            indextype=(PerperTypeUtils.get_java_type_name(type_) or type_.name)
            if ((flags and StreamFlags.query) != 0 and type_ != None)
            else None,
            indexfields=(
                1,
                {},
            ),  # self.serializer.get_queriable_fields(type) if (flags & self.stream_flags.query) != 0 and type != None else None,
            ephemeral=(flags and StreamFlags.ephemeral) != 0,
        )
        streams_cache.put(stream_name, stream_data)
        self._logger.debug(f"Created stream with name :{stream_name}")

    async def call(self, agent_name, agent_delegate, call_delegate, parameters):
        calls_cache = self.ignite.get_cache("calls")
        call_name = self.generate_name(call_delegate)
        call_data = CallData(
            agent=agent_name,
            agentdelegate=agent_delegate,
            delegate=call_delegate,
            calleragentdelegate=self.fabric.agent_delegate
            if self.fabric.agent_delegate != None
            else "",
            caller=self.instance.instance_name,
            finished=False,
            localtodata=True,
            error="",
            parameters=ParameterData(parameters=(1, parameters)),
        )
        calls_cache.put(call_name, call_data)

        self._logger.debug(f"Created call with name: {call_name}")
        key, notification = await self.fabric.get_call_notification(call_name)
        self.fabric.consume_notification(key)

        call = calls_cache.get_and_remove(notification.call)

        if call.error != "":
            raise Exception(f"Exception in call to {call_delegate}: {call.error}")
        else:
            return call

    def generate_name(self, basename=None):
        return f"{basename}-{uuid.uuid4()}"
